### ISSUE TITLE
ADK Core: Dollar Cost Estimation Data

### ISSUE THREAD

**Issue by 3p3r** [zorse-lang/adk#27]
Cost estimation data in ADK is offered on a best-effort basis and they're snapshots.
Their purpose is rough budgeting and not accurate accounting.

Cost estimation data of a Component is modeled as follows:

```ts
interface Cost {
  // USD cost of Component once snapped into a Scene.
  readonly usd: number;
  // Payment has terms, e.g. "pay-as-you-go", "upfront", "monthly", "yearly", etc.
  readonly terms: string[] | undefined;
  // Charge is either flat or variable. once variable, it has some conditions. e.g. "between 1k and 5k calls", "after 1M requests" etc.
  // Conditions are meant to be scoped down to a Component type
  readonly conditions: string[] | undefined;
}
```

`Cost`s can be added up together with the following logic:

```mermaid
graph LR
  A{terms match?}
  B{conditions match?}
  C{component types match?}
  E{both conditions empty?}
  *((can we add up?))
  X((end))
  Y[we can add up]
  N[can't add up]

  * --> B
  B --> |y| E --> |y| A --> |y| Y
  E --> |n| C
  B --> |n| N
  C --> |y| A --> |n| N
  Y --> X
  N --> X
```

Every `Component` has an optional cost vector associated with it: `readonly cost: Cost[][] | undefined`.

Lack of a cost vector means there were no cost data available for the `Component` during generation or its cost data is unknown.
> Note that this is different from `cost` being an empty vector which means this resource is free $0 (e.g local files, local k8s, etc.).

`cost` is an array of array of `Cost`s. First level indicates different "offerings".  
Second level is multiple types of charges incurred in the same "offering" (e.g if a resource costs both monthly and conditionally after).

In other words, in the 2D matrix of costs, moving vertically is equivalent to a logical OR in a quote and going horizontally is equivalent to a logical AND in a quote from a cloud billing provider.

Retrieving this `cost` vector must be `Scene` dependent to support multiple regions for the same `Component` type. Something like:

```ts
Scene.cost(component: Component): Cost[][] | undefined; // cost of a single Component in this Scene
Scene.cost(): Cost[][] | undefined; // aggregate cost of the entire Scene
```

At the end of the render pipeline, `Scene.cost` is called to aggregate the data and report to user.

Currency must be converted to USD before putting into the `usd` field.
